mercury_engine_data_structures/formats/bsmat.py

```diff
@@ -0,0 +1,246 @@
+)
+
+from mercury_engine_data_structures.common_types import Char, Float, StrId
+from mercury_engine_data_structures.formats import BaseResource
+from mercury_engine_data_structures.game_check import Game
+
+# these seem to be using Unity ShaderLab, or at least the gist I borrowed this from matches up
+# source for most binary info: https://gist.github.com/KillzXGaming/9817455559544cb3613f99184aa3ed68
+
+# === ENUMS ===
+
+# seems to be missing some values, example: missile expansion material uses 8
+# cant seem to find info on google easily so no docs
+translucency_type = Enum(
+    Int32ul,
+    TILING_CLAMP=0,
+    TILING_CLAMPCOLOR=1,
+    TILING_REPEAT=2,
+    TILING_MIRROR=3,
+    TILING_INVALID=0xffffffff
+)
+
+# blend operation
+# https://docs.unity3d.com/Manual/SL-BlendOp.html
+blend_op = Enum(
+    Int32ul,
+    ADD=0,
+    SUB=1,
+    REV_SUB=2,
+    MIN=3,
+    MAX=4,
+    INVALID=0xffffffff
+)
+
+# blend mode
+# https://docs.unity3d.com/Manual/SL-Blend.html
+blend_mode = Enum(
+    Int32ul,
+    ZERO=0,
+    ONE=1,
+    SRC_COLOR=2,
+    ONE_MINUS_SRC_COLOR=3,
+    DST_COLOR=4,
+    ONE_MINUS_DST_COLOR=5,
+    SRC_ALPHA=6,
+    ONE_MINUS_SRC_ALPHA=7,
+    DST_ALPHA=8,
+    ONE_MINUS_DST_ALPHA=9,
+    INVALID=0xffffffff
+)
+
+# cull command
+# https://docs.unity3d.com/Manual/SL-Cull.html
+polygon_cull_mode = Enum(
+    Int32ul,
+    BACK = 2,
+    FRONT = 3,
+    OFF = 4,
+    INVALID = 0xffffffff
+)
+
+# stencil operation value
+# https://docs.unity3d.com/Manual/SL-Stencil.html#stencil-operation-values
+stencil_op = Enum(
+    Int32ul,
+    KEEP=0,
+    ZERO=1,
+    REPLACE=2,
+    INCR_SAT=3,
+    DECR_SAT=4,
+    INVERT=5,
+    INCR_WRAP=6,
+    DECR_WRAP=7,
+    INVALID=0xffffffff
+)
+
+# idk
+compare_mode = Enum(
+    Int32ul,
+    CMPMODE_ALWAYS=0,
+    CMPMODE_NEVER=1,
+    CMPMODE_EQUAL=2,
+    CMPMODE_NOTEQUAL=3,
+    CMPMODE_LESS=4,
+    CMPMODE_LESSEQUAL=5,
+    CMPMODE_GREATER=6,
+    CMPMODE_GREATEREQUAL=7,
+    CMPMODE_MAX_COUNT=8,
+    CMPMODE_INVALID=0xffffffff
+)
+
+# idk either
+fill_mode = Enum(
+    Int32ul,
+    SOLID=0,
+    WIRE=1,
+    INVALID=0xffffffff
+)
+
+shader_type = Enum(
+    Int32ul,
+    VERTEX=0,
+    PIXEL=1,
+    GEOMETRY=2,
+    INVALID=0xffffffff
+)
+
+# seems to be a fancier version of point/bilinear/trilinear with extra trilinear/mipmap stuff
+filter_mode = Enum(
+    Int32ul,
+    FILTER_NEAREST=0,
+    FILTER_LINEAR=1,
+    FILTER_NEAREST_MIP_NEAREST=2,
+    FILTER_NEAREST_MP_LINEAR=3,
+    FILTER_LINEAR_MIP_NEAREST=4,
+    FILTER_LINEAR_MIP_LINEAR=5,
+    FILTER_INVALID=0xffffffff
+)
+
+tile_wrap_mode = Enum(
+    Int32ul,
+    TILING_CLAMP=0,
+    TILING_CLAMP_COLOR=1,
+    TILING_REPEAT=2,
+    TILING_MIRROR=3,
+    TILING_INVALID=0xffffffff
+)
+
+# === STRUCTS ===
+
+# seems to be a combo of Blend and BlendOp
+# ie "Blend source dest && BlendOp operation"
+blend_state = Struct(
+    enabled = Byte,
```

Comment:
Construct has a `Flag` type which is Byte, but converts to True/False. Does it fit?

Reply:
Not sure, but probably. I just used byte to make sure that if it isn't it will still build a matching version.

Reply:
Seems like a good case to add test cases! Check `test_bmsad.py` for example, doing something similar for this would be great.